internal/ethapi/api.go

```diff
@@ -1208,11 +1208,16 @@ func applyMessage(ctx context.Context, b Backend, args TransactionArgs, state *s
 	if precompiles != nil {
 		evm.SetPrecompiles(precompiles)
 	}
-
-	return applyMessageWithEVM(ctx, evm, msg, state, timeout, gp)
+	res, err := applyMessageWithEVM(ctx, evm, msg, timeout, gp)
+	// If an internal state error occurred, let that have precedence. Otherwise,
+	// a "trie root missing" type of error will masquerade as e.g. "insufficient gas"
+	if err := state.Error(); err != nil {
```

Comment:
It's because the new code does not pass `state` along to `applyMessageWithEvm`. Originally, I wen ta bit back and forth between passing `*state.StateDB` and `vm.StateDB`, but in the end I realized that I don't have to even decide. But then I need to pull the error-check up here. 

> These two approaches should be totally equivalent? 

Yes, I hope so :)